### PR TITLE
Added unit tests for NaN, Max, Min, Positive and Negative Infinity values

### PR DESCRIPTION
**Pull request by atulagrawal**
Fixes # . https://github.com/cbeust/testng/issues/1734

### Did you remember to?

- [x] Add test case(s)
- [x] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

File: src/test/java/org/testng/AssertTest.java [hunk 1]
```diff
@@ -164,18 +164,32 @@ public void compareCharacterArrays() {
 
   @Test
   public void compareFloatArrays() {
-    float[] actual = {(float) Math.PI, (float) Math.E};
-    float[] expected = {(float) Math.PI, (float) Math.E};
+    float[] actual = {(float) Math.PI, (float) Math.E, Float.MIN_VALUE, Float.MIN_NORMAL, Float.MAX_VALUE, Float.POSITIVE_INFINITY,
+            Float .NEGATIVE_INFINITY};
```
Comment (juherr) [inline src/test/java/org/testng/AssertTest.java]: You have an extra space here

File: src/test/java/org/testng/AssertTest.java [hunk 2]
```diff
@@ -164,18 +164,32 @@ public void compareCharacterArrays() {
 
   @Test
   public void compareFloatArrays() {
-    float[] actual = {(float) Math.PI, (float) Math.E};
-    float[] expected = {(float) Math.PI, (float) Math.E};
+    float[] actual = {(float) Math.PI, (float) Math.E, Float.MIN_VALUE, Float.MIN_NORMAL, Float.MAX_VALUE, Float.POSITIVE_INFINITY,
+            Float .NEGATIVE_INFINITY};
+    float[] expected = {(float) Math.PI, (float) Math.E,Float.MIN_VALUE, Float.MIN_NORMAL, Float.MAX_VALUE, Float.POSITIVE_INFINITY,
```
Comment (juherr) [inline src/test/java/org/testng/AssertTest.java]: And a space is missing here

File: src/test/java/org/testng/AssertTest.java [hunk 3]
```diff
@@ -164,18 +164,32 @@ public void compareCharacterArrays() {
...
-    float[] actual = {(float) Math.PI, (float) Math.E};
-    float[] expected = {(float) Math.PI, (float) Math.E};
+    float[] actual = {(float) Math.PI, (float) Math.E, Float.MIN_VALUE, Float.MIN_NORMAL, Float.MAX_VALUE, Float.POSITIVE_INFINITY,
+            Float .NEGATIVE_INFINITY};
+    float[] expected = {(float) Math.PI, (float) Math.E,Float.MIN_VALUE, Float.MIN_NORMAL, Float.MAX_VALUE, Float.POSITIVE_INFINITY,
+            Float.NEGATIVE_INFINITY};
     Assert.assertEquals(actual, expected);
   }
 
+  @Test(expectedExceptions = AssertionError.class)
```
Comment (juherr) [inline src/test/java/org/testng/AssertTest.java]: It is not the exected behavior from #1734. He'd like to not have an exception.

**Review by juherr — CHANGES_REQUESTED** [review]

**Comment by krmahadevan** [comment]
On a side, please have your PR rebase off of master. I have pushed in a lot
of code formatting changes to master. Just trying to ensure that we don't
run into merge conflicts. --

Thanks & Regards
Krishnan Mahadevan

"All the desirable things in life are either illegal, expensive, fattening
or in love with someone else!"
My Scribblings @ http://wakened-cognition.blogspot.com/

My Technical Scribbings @ http://rationaleemotions.wordpress.com/

**Comment by atulagrawal** [comment]
One more question - why we don't have public static void assertEquals(float actual, float expected, String message) method implementation? Same goes for double as well?

File: src/test/java/org/testng/AssertTest.java [hunk 4]
```diff
@@ -164,18 +164,41 @@ public void compareCharacterArrays() {
...
+    float[] expected = {(float) Math.PI, (float) Math.E, Float.MIN_VALUE, Float.MIN_NORMAL, Float.MAX_VALUE,
+            Float.POSITIVE_INFINITY, Float.NEGATIVE_INFINITY};
     Assert.assertEquals(actual, expected);
   }
 
+  @Test(expectedExceptions = AssertionError.class)
+  public void compareFloatArraysWithNaNValues() {
+    Assert.assertEquals(new float[] { Float.NaN }, new float[] { Float.NaN });
+  }
+  @Test(expectedExceptions = AssertionError.class)
```
Comment (juherr) [inline src/test/java/org/testng/AssertTest.java]: A space is missing between methods

**Comment by juherr** [comment]
> i see that the behavior for NaN value is different for float and double.

We definitely need to fix it.

> Please advise here and I will do the implementation accordingly.

`compareFloatWithNaNValues` must pass without the exception.
 
And I think array tests like `compareDoubleArraysWithNaNValues` should pass without exception too (WDYT @krmahadevan )

> why we don't have public static void assertEquals(float actual, float expected, String message) method implementation? Same goes for double as well?

Because no one added it. Feel free to add the method and its variants.

**Comment by atulagrawal** [comment]
Thanks @juherr. I have two choices here for fixing above mentioned differences:
1. Covert float to Float Object and compare them using equals method.
2. Use float as primitive only and use "==" operator but have special handling for NaN and Infinity scenarios. Something similar to 
```
public static void assertEquals(double actual, double expected, double delta, String message) {
    // handle infinity specially since subtracting to infinite values gives NaN and the
    // the following test fails
    if (Double.isInfinite(expected)) {
      if (!(expected == actual)) {
        failNotEquals(actual, expected, message);
      }
    } else if (Double.isNaN(expected)) {
      if (!Double.isNaN(actual)) {
        failNotEquals(actual, expected, message);
      }
    } else if (!(Math.abs(expected - actual) <= delta)) {
      failNotEquals(actual, expected, message);
    }
  }
```
The trade off here is between performance and readability. Option 1 will result in more object creations and may cause slight performance issues for very large arrays.

**Comment by krmahadevan** [comment]
@juherr - I haven't had time to go through this PR and the discussion yet since I was stuck with the formatting changes (I didn't anticipate it to be tiring 😛 ). 
So let me go through the conversation first and then chime in if required with inputs.

**Comment by juherr** [comment]
@atulagrawal I prefer option 2

**Comment by atulagrawal** [comment]
@juherr - I have done the changes as per option 2. Can you please review it.

File: src/main/java/org/testng/Assert.java [hunk 5]
```diff
@@ -369,16 +369,13 @@ public static void assertEquals(float[] actual, float[] expected, String message
...
     for (int i = 0; i < expected.length; i++) {
-      if (expected[i] != actual[i]) {
-        fail(
-            String.format(
-                ARRAY_MISMATCH_TEMPLATE,
-                i,
-                Float.toString(expected[i]),
-                Float.toString(actual[i]),
-                message));
+      assertEquals(actual[i],expected[i], String.format(
```
Comment (juherr) [inline src/main/java/org/testng/Assert.java]: space missing after `,`

**Review by juherr — APPROVED** [review]
LGTM 👍 

Ping @cbeust

File: src/main/java/org/testng/Assert.java [hunk 6]
```diff
@@ -369,16 +369,13 @@ public static void assertEquals(float[] actual, float[] expected, String message
...
-      if (expected[i] != actual[i]) {
-        fail(
-            String.format(
-                ARRAY_MISMATCH_TEMPLATE,
-                i,
-                Float.toString(expected[i]),
-                Float.toString(actual[i]),
-                message));
+      assertEquals(actual[i], expected[i], String.format(
+                                              ARRAY_MISMATCH_TEMPLATE,
```
Comment (cbeust) [inline src/main/java/org/testng/Assert.java]: Not sure what the logic for this formatting is, it should just be indented one or possibly two from the previous line.

**Review by cbeust — COMMENTED** [review]
Looks good to me after the formatting is fixed.

**Comment by atulagrawal** [comment]
@cbeust - is it fine now?

**Pull request merged**